datafusion/physical-expr/src/simplifier/unwrap_cast.rs

```diff
@@ -0,0 +1,621 @@
+    expr.transform_down(|e| {
+        if let Some(binary) = e.as_any().downcast_ref::<BinaryExpr>() {
+            if let Some(unwrapped) = try_unwrap_cast_binary(binary, schema)? {
+                return Ok(Transformed::yes(unwrapped));
+            }
+        }
+        Ok(Transformed::no(e))
+    })
+}
+
+/// Try to unwrap casts in binary expressions
+fn try_unwrap_cast_binary(
+    binary: &BinaryExpr,
+    schema: &Schema,
+) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+    // Case 1: cast(left_expr) op literal
+    if let (Some((inner_expr, _cast_type)), Some(literal)) = (
+        extract_cast_info(binary.left()),
+        binary.right().as_any().downcast_ref::<Literal>(),
+    ) {
+        if let Some(unwrapped) = try_unwrap_cast_comparison(
+            Arc::clone(inner_expr),
+            literal.value(),
+            *binary.op(),
+            schema,
+        )? {
+            return Ok(Some(unwrapped));
+        }
+    }
+
+    // Case 2: literal op cast(right_expr)
+    if let (Some(literal), Some((inner_expr, _cast_type))) = (
+        binary.left().as_any().downcast_ref::<Literal>(),
+        extract_cast_info(binary.right()),
+    ) {
+        // For literal op cast(expr), we need to swap the operator
+        let swapped_op = swap_operator(*binary.op());
+        if let Some(unwrapped) = try_unwrap_cast_comparison(
+            Arc::clone(inner_expr),
+            literal.value(),
+            swapped_op,
+            schema,
+        )? {
+            // unwrapped is already the correct expression: inner_expr swapped_op casted_literal
+            return Ok(Some(unwrapped));
+        }
+    }
+
+    Ok(None)
+}
+
+/// Extract cast information from a physical expression
+fn extract_cast_info(
+    expr: &Arc<dyn PhysicalExpr>,
+) -> Option<(&Arc<dyn PhysicalExpr>, &DataType)> {
+    if let Some(cast) = expr.as_any().downcast_ref::<CastExpr>() {
+        Some((cast.expr(), cast.cast_type()))
+    } else if let Some(try_cast) = expr.as_any().downcast_ref::<TryCastExpr>() {
+        Some((try_cast.expr(), try_cast.cast_type()))
+    } else {
+        None
+    }
+}
+
+/// Try to unwrap a cast in comparison by moving the cast to the literal
+fn try_unwrap_cast_comparison(
+    inner_expr: Arc<dyn PhysicalExpr>,
+    literal_value: &ScalarValue,
+    op: Operator,
+    schema: &Schema,
+) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+    // Get the data type of the inner expression
+    let inner_type = inner_expr.data_type(schema)?;
+
+    // Try to cast the literal to the inner expression's type
+    if let Some(casted_literal) = try_cast_literal_to_type(literal_value, &inner_type) {
+        let literal_expr = lit(casted_literal);
+        let binary_expr = BinaryExpr::new(inner_expr, op, literal_expr);
+        return Ok(Some(Arc::new(binary_expr)));
+    }
+
+    Ok(None)
+}
+
+/// Swap comparison operators for right-side cast unwrapping
+fn swap_operator(op: Operator) -> Operator {
```

Comment:
This could use `Operator::swap`, right? I don't think there is any reason to have it specially here